### PR TITLE
Add BIR spec and test suite

### LINKED ISSUE #24771: Document BIR format
Create and publish BIR format spec at https://github.com/ballerina-platform/ballerina-lang/tree/master/docs/compiler

### LINKED ISSUE #25532: Merge BIR spec changes from nballerina branch to master
**Description:**
The BIR spec changes worked on nballerina branch can be pushed to master as well since its generic and master should also have BIR validation in place.

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### LINKED ISSUE #25535: Add a comprehensive test case to cover BIR spec validation completely
**Description:**
Currently, the spec validation test suite has few isolated tests but having a single comprehensive test to cover all aspects would be an ideal thing.

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by Kishanthan**
## Purpose
This PR adds BIR specification as a binary representation format using the Kaitai Struct Language (https://doc.kaitai.io/user_guide.html).

It also includes a test suite that validates BIR spec against common ballerina constructs using few ballerina source files.

Fixes #24771 
Fixes #25532
Fixes #25535

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: docs/bir-spec/spotbugs-exclude.xml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<!--
+  ~  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (riyafa) [inline docs/bir-spec/spotbugs-exclude.xml:2]: Shall we update the license header?

File: docs/bir-spec/src/main/java/org/ballerinalang/birspec/BIRSpecGenerator.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package org.ballerinalang.birspec;
```
Comment (riyafa) [inline docs/bir-spec/src/main/java/org/ballerinalang/birspec/BIRSpecGenerator.java:1]: Shall we add the license header?

File: docs/bir-spec/spotbugs-exclude.xml (hunk 1 above)
Comment (Kishanthan) [reply docs/bir-spec/spotbugs-exclude.xml]: ack

File: docs/bir-spec/src/main/java/org/ballerinalang/birspec/BIRSpecGenerator.java (hunk 2 above)
Comment (Kishanthan) [reply docs/bir-spec/src/main/java/org/ballerinalang/birspec/BIRSpecGenerator.java:18]: done

**Review by riyafa — APPROVED** [review]

**Pull request merged**